symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/reflection/system/ListLinePlot3D.java

```diff
@@ -29,140 +29,154 @@ public IExpr evaluate(final IAST ast, EvalEngine engine) {
-          IASTAppendable result = F.Graphics3D(temp);
-          if (ast.argSize() > 1) {
-            // add same options to Graphics3D
-            result.appendAll(ast, 2, ast.size());
-          }
-          return result;
+      // case 2: single line coordinates
+      // e.g.: ListLinePlot3D[{{x_1, y_1, z_1}, {x_2, y_2, z_2}}]
+      if (dimension != null && dimension.length == 2 && dimension[1] == 3) {
+        IASTAppendable result = F.Graphics3D(coordinateLinePlot((IAST) F.List(ast.arg1()), plotStyle, engine));
+        if (ast.argSize() > 1) {
+          // add same options to Graphics3D
+          result.appendAll(ast, 2, ast.size());
         }
-        return F.NIL;
+        return result;
       }
 
-      // second try if arg1 is a list of matrices
-      if (ast.arg1().isList()) {
-        IAST list = (IAST) ast.arg1();
-        IASTAppendable resultList = F.ListAlloc(list.size());
-        double deltaHeight = 1.0;
-        for (int i = 1; i < list.size(); i++) {
-          int[] dims = list.get(i).isMatrix(false);
-          if (dims != null && dims.length == 2) {
-            IExpr flattenHeights = engine.evaluate(F.Flatten(list.get(i)));
-            double tempHeight =
-                engine.evaluate(F.Max(flattenHeights).subtract(F.Min(flattenHeights))).evalDouble();
-            deltaHeight = Math.max(deltaHeight, tempHeight);
-          }
-        }
-        for (int i = 1; i < list.size(); i++) {
-          int[] dims = list.get(i).isMatrix(false);
-          if (dims != null && dims.length == 2) {
-            IExpr temp =
-                matrixLinePlot(list.get(i), dims, plotStyle, lineColorNumber, deltaHeight, engine);
-            if (temp.isPresent()) {
-              resultList.append(temp);
-            }
-          }
+      // case 3: multiple line heights
+      // e.g.: ListLinePlot3D[{{1, 2, 3, 4}, {-1, -2, -3, -4}}]
+      if (
+        ast.arg1().isList() &&
+        ((IAST) ast.arg1()).arg1().isList() &&
+        ((IAST) ((IAST) ast.arg1()).arg1()).arg1().isNumber()
+      ) {
+        IASTAppendable result = F.Graphics3D(heightLinePlot((IAST) ast.arg1(), plotStyle, engine));
+        if (ast.argSize() > 1) {
+          // add same options to Graphics3D
+          result.appendAll(ast, 2, ast.size());
         }
-        IASTAppendable result = F.Graphics3D(resultList);
+        return result;
+      }
+
+      dimension = ((IAST) ast.arg1()).arg1().isMatrix(false);
+      // case 4: multiple line coordinates
+      // e.g.: ListLinePlot3D[{{coord1, coord2}, {coord3, coord4}}]
+      if (dimension != null && dimension.length == 2 && dimension[1] == 3) {
+        IASTAppendable result = F.Graphics3D(coordinateLinePlot((IAST) ast.arg1(), plotStyle, engine));
         if (ast.argSize() > 1) {
           // add same options to Graphics3D
           result.appendAll(ast, 2, ast.size());
         }
         return result;
       }
-    }
 
+      // the code reach here, NIL is returned
```

Comment:
With commit 6323f17 I added the `ldata` message, which you can return in the following snippet if the input argument of `ListPlot3D` isn't correct:

```java
// `1` is not a valid dataset or a list of datasets.
      return IOFunctions.printMessage(ast.topHead(), "ldata", F.List(ast.arg1()), engine);
```

Reply:
Done.